apps/shipments/serializers/shipment.py

```diff
@@ -184,7 +188,30 @@ def create(self, validated_data):
             if 'device' in self.context:
                 extra_args['device'] = self.context['device']
 
-            return Shipment.objects.create(**validated_data, **extra_args)
+            if 'tags' in validated_data:
+                shipment_tags = validated_data.pop('tags')
+                shipment = Shipment.objects.create(**validated_data, **extra_args)
+                for shipment_tag in shipment_tags:
+                    ShipmentTag.objects.create(
+                        tag_type=shipment_tag['tag_type'],
+                        tag_value=shipment_tag['tag_value'],
+                        shipment_id=shipment.id,
+                        owner_id=shipment.owner_id
+                    )
+            else:
+                shipment = Shipment.objects.create(**validated_data, **extra_args)
+
+            return shipment
+
+    def validate_tags(self, tags):
+        if len(tags) != len([set(t) for t in {tuple(d.items()) for d in tags}]):
```

Comment:
This is creative, but is not very readable. It also looks like you're taking the len of a list of sets? I think maybe you meant to convert the list of tuples to a set instead of converting each tuple into a set before putting it into the list, but you've already used a set comprehension ({}) so you have a list of unique tags with just this:
`{tuple(tag.items()) for tag in tags}`